src/Pomelo.EntityFrameworkCore.MySql/Storage/SynchronizedMySqlDataReader.cs

```diff
@@ -1,199 +1,200 @@
-            {
-                CloseReader();
-            }
-            return result;
-        }
-
-        public override T GetFieldValue<T>(int ordinal)
-        {
-            try
-            {
-                // try normal casting
-                return GetReader().GetFieldValue<T>(ordinal);
-            }
-            catch (InvalidCastException e)
-            {
-                return ConvertWithReflection<T>(ordinal, e);
-            }
-        }
-
-        public override async Task<T> GetFieldValueAsync<T>(int ordinal, CancellationToken cancellationToken)
-        {
-            try
-            {
-                // try normal casting
-                return await GetReader().GetFieldValueAsync<T>(ordinal, cancellationToken).ConfigureAwait(false);
-            }
-            catch (InvalidCastException e)
-            {
-                return ConvertWithReflection<T>(ordinal, e);
-            }
-        }
-
-        private T ConvertWithReflection<T>(int ordinal, InvalidCastException e)
-        {
-            try
-            {
-                // try casting using reflection; needed for json
-                var dataParam = Expression.Parameter(typeof(string), "data");
-                var body = Expression.Block(Expression.Convert(dataParam, typeof(T)));
-                var run = Expression.Lambda(body, dataParam).Compile();
-                return (T) run.DynamicInvoke(GetReader().GetValue(ordinal));
-            }
-            catch (Exception)
-            {
-                // throw original InvalidCastException
-                throw e;
-            }
-        }
-
-        private void CloseReader()
-        {
-            try
-            {
-                if (_reader != null)
-                {
-                    // dispose the underlying MySQL data reader
-                    _reader.Dispose();
-                    _reader = null;
-                }
-            }
-            finally
-            {
-                if (_lock != null)
-                {
-                    // release the shared lock, so another statement can be executed
-                    _lock.Release();
-                    _lock = null;
-                }
-            }
-        }
-
-        protected override void Dispose(bool disposing)
-        {
-            CloseReader();
-            base.Dispose(disposing);
-        }
-
-        private DbDataReader GetReader()
-        {
-            if (_reader == null)
-                throw new ObjectDisposedException(nameof(SynchronizedMySqlDataReader));
-            return _reader;
-        }
-    }
-}
+﻿using System;
+using System.Collections;
+using System.Data.Common;
+using System.IO;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore.Storage.Internal;
+using MySql.Data.MySqlClient;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Storage
```

Comment:
@Kagamine I moved it from `Microsoft.EntityFrameworkCore.Storage.Interna`l to `Microsoft.EntityFrameworkCore.Storage`.  The Internal namespaces are typically for objects that can be changed out through Dependency Injection.  I see Microsoft using the parent namespace (`Microsoft.EntityFrameworkCore.Storage`) in this case for implementation that cannot be changed out through DI.

Let's stay out of the `MySqlConnector` namespace.  This Synchronized Reader could technically synchronize any `DbDataReader`, and has a pretty tight coupling to the lock in `Microsoft.EntityFrameworkCore.Storage.Internal.MySqlRelationalConnection`.